internal/providers/provider.go

```diff
@@ -237,6 +237,42 @@ type ReadResourceRequest struct {
 	// each provider, and it should not be used without coordination with
 	// HashiCorp. It is considered experimental and subject to change.
 	ProviderMeta cty.Value
+
+	// DeferralAllowed signals that the provider is allowed to defer the
+	// changes. If set the caller needs to handle the deferred response.
+	DeferralAllowed bool
+}
+
+type DeferredReason string
```

Comment:
nit: maybe add a comment explaining the difference between the protocol and what we support here. Just clarify that we have additional reasons that can be generated in Terraform Core only, and it's easier just to keep the definition together in one place.